[READY] Lighting list cleanup.

MrsTonedOne opened the pull request:
Removed a bunch of back reference lists that were either entirely unused, or contained references that could be found within the datums contained in other lists.

Lighting corner datums now get deleted when unused.

Light sources no longer track lighting corners where the applied light rounds to 0.

Lighting corners are no longer tracked by a constant 4 item list. This was better as 4 datum vars.

~~todo: test. This is a branch from November, iirc it was testing fine, gonna do that now that i've rebased.~~ tested fine

edit:

Lighting corners are now lazy inited and deleted. they should always (and only exist) if there is a light source shining on it within range (even if the turf has no dynamic lighting). This is needed to support turfs that become lighting enabled mid round. On the plus side, they will no longer be generated on full dark turfs.

MrsTonedOne committed:
Lighting list cleanup.

Removed a bunch of back reference lists that were either entirely unused, or contained references that could be found within the datums contained in other lists.

lighting corner datums now get deleted with unused.

light sources no longer track lighting corners where the appiled light rounds to 0.

MrStonedOne commented:
Tested, spaming fix/break lights produces no runtimes or observable bugs, spaming shuttle moves in light interesting areas produces no runtimes or observable bugs.

Test merging on terry since it could use the extra memory.

MrStonedOne commented on code/modules/lighting/lighting_source.dm:
```diff
@@ -103,38 +102,39 @@
+		} else {                                 \
+			effect_str -= C                      \
+		}                                        \
+    	                                         \
```
```suggestion
		                                           \
```

MrStonedOne commented on code/modules/lighting/lighting_turf.dm:
```diff
@@ -122,15 +112,28 @@
+		if (recalc)
+			lighting_corner_NW.full_update()
+	
+	lighting_corners_initialised = TRUE
```
```suggestion
	lighting_corners_initialised = TRUE

```

MrsTonedOne committed:
Fix lint

MrStonedOne commented:
lighting init time is now 20% faster.

![image](https://user-images.githubusercontent.com/7069733/117589862-1b795a00-b0e1-11eb-8100-7cada1466398.png)

Kylerace commented on code/modules/lighting/lighting_corner.dm:
```diff
@@ -39,49 +40,55 @@ GLOBAL_LIST_INIT(LIGHTING_CORNER_DIAGONAL, list(NORTHEAST, SOUTHEAST, SOUTHWEST,
-			active = TRUE
-			return
+	var/datum/light_source/L
+	for (thing in affecting)
```
as anything

Kylerace commented on code/modules/lighting/lighting_source.dm:
```diff
@@ -231,50 +231,34 @@
 	LAZYINITLIST(effect_str)
 	if (needs_update == LIGHTING_VIS_UPDATE)
-		for (thing in  corners - effect_str) // New corners
+		for (thing in corners - effect_str) // New corners
```
all the loops that use thing can instead use as anything

MrsTonedOne committed:
cleanup for loops and fix bugs with negitive light sources

Cyberboss commented on code/modules/lighting/lighting_corner.dm:
```diff
@@ -134,6 +150,24 @@ GLOBAL_LIST_INIT(LIGHTING_CORNER_DIAGONAL, list(NORTHEAST, SOUTHEAST, SOUTHWEST,
 	if (!force)
 		return QDEL_HINT_LETMELIVE
 
-	stack_trace("Ok, Look, /tg/, I need you to find whatever fucker decided to call qdel on a fucking lighting corner, then tell him very nicely and politely that he is brain damaged and needs his head checked. Thanks. Send them my regards by the way.")
```
This is vintage PJB salt. Shame to lose it

LemonInTheDark commented:
I'm currently tearing my hair out over a harddel caused by lighting datums being held inside lava's `affecting_lights` list. I'm just gonna pray this fixes it and move on

optimumtact commented:
Why was this drafted it's been on basil for like weeks

MrStonedOne commented:
> Why was this drafted it's been on basil for like weeks

breaks on new construction of lighting enabled tiles (ie, space -> floor tile)

MrsTonedOne committed:
Fix lighting on turfs that gained dynamic lighting mid round. lazy init corner datums.

these two are related. by decoupling corner datums from the turfs dynamic lighting state, we can use them to know what level of light a non-dynamic light turf should have once it gains dynamic light.

Also should free up some memory not storing these datums in maint. Corner datums only exist on a turf that has light cast upon it by the dynamic lighting system.

MrStonedOne commented:
> I'm currently tearing my hair out over a harddel caused by lighting datums being held inside lava's `affecting_lights` list. I'm just gonna pray this fixes it and move on

this pr removes that list.

LemonInTheDark commented on code/modules/lighting/lighting_corner.dm:
```diff
@@ -39,49 +40,48 @@ GLOBAL_LIST_INIT(LIGHTING_CORNER_DIAGONAL, list(NORTHEAST, SOUTHEAST, SOUTHWEST,
-		masters[T]   = ((T.x > x) ? EAST : WEST) | ((T.y > y) ? NORTH : SOUTH) // Get the dir based on coordinates.
-		i            = GLOB.LIGHTING_CORNER_DIAGONAL.Find(turn(masters[T], 180))
-		T.corners[i] = src
+		save_master(T, ((T.x > x) ? EAST : WEST) | ((T.y > y) ? NORTH : SOUTH)) // Get the dir based on coordinates.
```
Would it be worth making this a define?

MrStonedOne replied:
My mind will melt if i stare at this code any longer, you can pr it to my branch or pr when you merge this.

Its a target for sure, now that corner datums get created later in the process and self clean up when lights go away.

LemonInTheDark replied:
aight, that's a good enough reason for me

LemonInTheDark commented:
Lighting objects are hard deleting real bad, but not very often. I'm unsure of what could be causing this, but mso seemed to have some idea. <https://tgstation13.org/parsed-logs/terry/data/logs/2021/05/28/round-163093/runtime.log>
It's not common, but when it fires it's really bad. Wish I had the chance to sic the ref tracker on this

YakumoChen commented:
Replacing tiles seems to fuck with the lighting in a way that triggers Nyctophobia even if you're in a lit area, I assume it's related.

LemonInTheDark commented:
What server was this happening on?
